doc/neps/nep-0018-array-function-protocol.rst

```diff
@@ -93,24 +93,21 @@ We propose the following signature for implementations of
 
 -  ``func`` is an arbitrary callable exposed by NumPy's public API,
    which was called in the form ``func(*args, **kwargs)``.
--  ``types`` is a list of argument types from the original NumPy
-   function call that implement ``__array_function__``, in the order in which
-   they will be called.
+-  ``types`` is a list of unique argument types from the original NumPy
+   function call that implement ``__array_function__``.
 -  The tuple ``args`` and dict ``kwargs`` are directly passed on from the
    original call.
 
 Unlike ``__array_ufunc__``, there are no high-level guarantees about the
 type of ``func``, or about which of ``args`` and ``kwargs`` may contain objects
-implementing the array API. As a convenience for ``__array_function__``
-implementors, ``types`` contains a list of argument types with an
-``'__array_function__'`` attribute. This allows downstream implementations to
-quickly determine if they are likely able to support the operation.
+implementing the array API.
 
-Still be determined: what guarantees can we offer for ``types``? Should
-we promise that types are unique, and appear in the order in which they
-are checked? Should we pass in arguments directly instead, either the full
-list of arguments in ``relevant_arguments`` (see below) or a single argument
-for each unique type?
+As a convenience for ``__array_function__`` implementors, ``types`` contains a
+list of argument types with an ``'__array_function__'`` attribute. This allows
+downstream implementations to quickly determine if they are likely able to
+support the operation. There are no guarantees about the order of argument types in ``types``: ``__array_funtion__`` implementations that rely on such behavior
```

Comment:
Should `types` be a set if we don't guarantee order? Would be easier to implement (no checking needed for whether a given class was already seen).

Reply:
I'm pretty sure that eventually we'll implement `try_array_function_override` in C, so we're definitely not making a `types` a Python set/list object until we're sure we need to call `__array_function__`. So convenient implementation isn't really a concern, but I agree that it's a better for our interface to enforce invariants, as long as we don't pay a performance penalty. And indeed, the speed difference for creating a small `set` vs a `list` vs a `tuple` is pretty small:
```
class A:
  pass

class B:
  pass

x = [A, B]
%timeit tuple(x)  # 158 ns per loop
%timeit list(x)  # 186 ns per loop
%timeit set(x)  # 225 ns per loop
```